hipparchus-optim/src/main/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizer.java

```diff
@@ -0,0 +1,340 @@
+ * the linearized problem at each iteration.
+ * </p>
+ *
+ * @author Julie Bayard
+ */
+public class SequentialGaussNewtonOptimizer implements LeastSquaresOptimizer {
+
+    /**
+     * The singularity threshold for matrix decompositions. Determines when a
+     * {@link MathIllegalStateException} is thrown. The current value was the
+     * default value for {@link LUDecomposition}.
+     */
+    private static final double SINGULARITY_THRESHOLD = 1e-11;
+
+    /** Decomposer. */
+    private final MatrixDecomposer decomposer;
+
+    /** Old evaluation previously computed. */
+    private final Evaluation oldEvaluation;
+
+    /** Old jacobian previously computed. */
+    private final RealMatrix oldLhs;
+
+    /** Old residuals previously computed. */
+    private final RealVector oldRhs;
+
+    /**
+     * Create a sequential Gauss Newton optimizer.
+     * <p>
+     * The default for the algorithm is to solve the normal equations
+     * J<sup>T</sup>Jx=J<sup>T</sup>r using QR decomposition.
+     * </p>
+     *
+     * @param evaluation old evaluation previously computed.
+     */
+    public SequentialGaussNewtonOptimizer(final Evaluation evaluation) {
+        this(new QRDecomposer(SINGULARITY_THRESHOLD), evaluation);
+    }
+
+    /**
+     * Create a sequential Gauss Newton optimizer that uses the given matrix
+     * decomposition algorithm to solve the normal equations.
+     * <p>
+     * The {@code decomposer} is used to solve J<sup>T</sup>Jx=J<sup>T</sup>r.
+     * </p>
+     *
+     * @param decomposer the decomposition algorithm to use.
+     * @param evaluation old evaluation previously computed.
+     */
+    public SequentialGaussNewtonOptimizer(final MatrixDecomposer decomposer,
+                                          final Evaluation evaluation) {
+        this.decomposer = decomposer;
+        this.oldEvaluation = evaluation;
+        final Pair<RealMatrix, RealVector> normalEquation =
+                        computeNormalMatrix(evaluation.getJacobian(), evaluation.getResiduals());
+                    // solve the linearized least squares problem
+        this.oldLhs = normalEquation.getFirst();
+        this.oldRhs = normalEquation.getSecond();
+    }
+
+    /**
+     * Get the matrix decomposition algorithm.
+     *
+     * @return the decomposition algorithm.
+     */
+    public MatrixDecomposer getDecomposer() {
+        return decomposer;
+    }
+
+    /**
+     * Get the previous evaluation used by the optimizer.
+     *
+     * @return the previous evaluation.
+     */
+    public Evaluation getOldEvaluation() {
+        return this.oldEvaluation;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Optimum optimize(final LeastSquaresProblem lsp) {
+        // create local evaluation and iteration counts
+        final Incrementor evaluationCounter = lsp.getEvaluationCounter();
+        final Incrementor iterationCounter = lsp.getIterationCounter();
```

Comment:
Are two different counters needed here? They seem to be incremented in parallel.